### ISSUE TITLE
Allow turning a few optimisations off for Kryo 5.4.0

### ISSUE THREAD

**Issue by chrisr3** [EsotericSoftware/kryo#943]
We are upgrading our code-base from Kryo 4.0.2 to 5.4.0, and have hit an issue when serialising a `Collection` only containing instances of `UnmodifiableRandomAccessList`.

We have sub-classed `Kryo` to support serialising classes with `writeResolve` and `readReplace` methods; `UnmodifiableRandomAccessList` implements `writeResolve` like this:
```java
private Object writeReplace() {
    return new UnmodifiableList<>(list);
}
```
Kryo 5's `CollectionSerializer` seems to have an optimisation for the case when all of collection's elements are of the same type, whereby Kryo serializes the element type once, and then serializes all of the elements' object data afterwards. Unfortunately, `UnmodifiableRandomAccessList.writeResolve` transforms the elements into instances of `UnmodifiableList`, but Kryo is unable to discover this until is has invoked `writeResolve` at least once. This is _after_ it has incorrectly written `UnmodifiableRandomAccessList` as the element type to the output stream, of course.

I cannot see a way of working around this apart from ignoring `UnmodifiableRandomAccessList.writeResolve`, although I cannot ignore _every_ `writeResolve` method because that would break lambda serialization. Ideally, I'd like to be able to configure `CollectionsSerializer` _not_ to apply this optimisation at all.

**Comment by theigl** [comment]
Hi @chrisr3. Could you create a minimal reproducer (ideally a unit test) for the issue you are facing?

I'm open to adding a flag to `CollectionSerializer`, but there might be other solutions.

**Comment by chrisr3** [comment]
This sample project should demonstrate what is happening:
[kryo-example.tar.gz](https://github.com/EsotericSoftware/kryo/files/11026870/kryo-example.tar.gz)

**Comment by theigl** [comment]
OK thanks! 

I quickly glanced over the code:

Do you really need to globally override Kryo's behavior? Or would it be enough to create a custom serializer that uses `writeResolve/readReplace` internally and register it only for the classes where you actually want this behavior?

Lambda serialization is handled by `ClosureSerializer` that uses `readResolve/writeReplace` internally. So lambda serialization won't break if you change your global serialization logic.

**Comment by chrisr3** [comment]
I think the problem is that _any_ class (including unknown "third party" ones) can (potentially) implement `readReplace` or `writeResolve`, so we cannot possibly know them all in advance.

**Comment by chrisr3** [comment]
> We could add a flag to `CollectionSerializer` that disables the optimization, but I'm not sure if any other Kryo user will ever need it.

Maybe a better and more general solution would be to allow us to override Kryo's built-in `CollectionSerializer` with one of our own?

**Comment by chrisr3** [comment]
Hmm, interesting. I wouldn't expect that to work because the built-in `CollectionSerializer` is already registered for `Collection.class` and so wouldn't `MyCollectionSerialzier` have lower precedence?

**Comment by theigl** [comment]
Default serializers registered after the built-in serializers should take priority. I just did a quick test and it looks like `MyCollectionSerializer` is chosen over `CollectionSerializer`. Please give it a try.

**Comment by chrisr3** [comment]
Yes, I agree. However, I have now discovered that my example code is actually testing the `ObjectArraySerializer` and not the `CollectionSerializer` :facepalm:. It looks like _both_ of these default serializers have this optimisation, and hence will need overriding. Are there any others please?

I have tried adding an alternate `CollectionSerializer`, but it seems that Kryo is still occasionally invoking the built-in one in some circumstances too.

**Comment by theigl** [comment]
> Yes, I agree. However, I have now discovered that my example code is actually testing the ObjectArraySerializer and not the CollectionSerializer 🤦. It looks like both of these default serializers have this optimisation, and hence will need overriding. Are there any others I need to worry about please?

I think these two are the only serializers that use this optimization.

> I have successfully added MyCollectionSerializer, but I also needed to re-add some other built-in serializers in order to prevent it from serializing all collection types:

OK this is unfortunate. It might sense to add something like `Kryo.replaceDefaultSerializer()`.

In the meantime, you could try this:

```java
Kryo kryo = new Kryo() {
    public Serializer getDefaultSerializer(Class type) {
        final Serializer serializer = super.getDefaultSerializer(type);
        if (serializer instanceof CollectionSerializer<?>) {
            return new MyCollectionSerializer();
        }
        return serializer;
    }
};
```

**Comment by chrisr3** [comment]
Oh dear, the `CollectionSerializer` has a few sub-classes!
- `ArraysAsListSerializer`
- `CopyForIterateCollectionSerializer`
- `JdkImmutableListSerializer`
- `JdkImmutableSetSerializer`
- `PriorityQueueSerializer`
- `TreeSetSerializer`

All of these will need extra handling too if I replace `CollectionSerializer` with `MyCollectionSerializer`... :construction_worker_man:.

**Comment by chrisr3** [comment]
I suppose the question becomes: Are there any ways that this `CollectionSerializerAdapter` can be made _less_ Evil? One way that occurs to me is that the individual `ImmutableCollectionSerializer` classes could be `public` instead of package private/. But the _real_ Dark Magic is how I need to invoke the `protected` methods.

**Comment by chrisr3** [comment]
> I just took another quick look and a possible point of contact could be `ClassResolver.register(Registration registration)`. Did you consider that method?

Heh, yes, but configuring serializers via the `ClassResolver` seemed like an odd thing to do... :wink:

> I have no issue with making the `ImmutableCollectionSerializer` classes public.

:+1:

> The protected methods in `CollectionSerializer` are a different topic.

That's OK, I wouldn't want to make them `public` either. I was just wondering if you could think of a nicer way than using `MethodHandle`s of getting the same result. Although curiously, the JVM seems _happy_ to invoke a `protected` method on a _sibling_ class via a `MethodHandle` :thinking:.
